### ISSUE TITLE
Feature request: predefined WAYLAND_DISPLAY and DISPLAY

### ISSUE THREAD

**Issue by muradm** [swaywm/sway#6439]
With xorg/i3 it was possible to run multiple separate/independent xorg-server/i3. xorg-server have the following related options:
```
$ X --help
Unrecognized option: --help
use: X [:<display>] [option]
<<SKIPPED>>
ttyxx                  server started from init on /dev/ttyxx
vtXX                   use the specified VT number
```
i3 has the following related options:
```
$ man i3
<<SKIPPED>>
ENVIRONMENT
   I3SOCK
       This variable overwrites the IPC socket path (placed in /tmp/i3-%u.XXXXXX/ipc-socket.%p by default, where %u is replaced with your UNIX username, %p is replaced with i3’s PID and XXXXXX is a string of random characters from the portable filename character set (see mkdtemp(3))). The IPC socket is used by external programs like i3-msg(1) or i3bar(1).
<<SKIPPED>>
```

Combination of these features, allowed to run multiple xorg-servers on different ttys.
- First one:
```
$ tty
/dev/tty2
$ X :2 tty2 vt2 # or similar
```
Then within Xsession:
```
I3SOCK=/run/user/1000/i3-2.sock
```
- Second one:
```
$ tty
/dev/tty3
$ X :3 tty3 vt3 # or similar
```
Then within Xsession:
```
I3SOCK=/run/user/1000/i3-3.sock
```
With the above, multiple instances of xorg-server and i3 would start. It could be either same user instances or different user instances.

With wayland/sway, right now it is impossible to achieve the same behavior.

There was a similar long standing feature request #4328, which was closed by [this](https://github.com/swaywm/sway/issues/4328#issuecomment-898861205).

Since wayland/sway architecture is different from xorg/i3, i would suggest to map behavior explained above as following:

- `SWAYSOCK` - already implemented, nothing to do here
- add `--wayland-display <display>` argument to sway
- add `--x-display <display>` argument to sway
- add `--tty <N>` - if specified which tty to start on, other wise current tty
- add `--vt <N>` - if specified which vt to start on, other wise current vt

`--wayland-display` and `--x-display` arguments will tell sway that user intends to run new wayland/sway instance at specified wayland display and X display (if X support enabled). In this way, sway should ignore any currently set WAYLAND_DISPLAY and DISPLAY environment variables (if set at all). Child processes should have WAYLAND_DISPLAY and DISPLAY set to values specified in these arguments.

`--tty` and `--vt` should be applicable only if `--wayland-display` and/or optionally `--x-display` are specified.

This way, wayland/sway will support existing behavior of xorg/i3.

sway should not check if values provided by `--wayland-display` and `--x-display` are actually used or not. It should just attempt to start new instance, in case of already existing instance, just fail with meaningful error.

**Comment by emersion** [comment]
I don't think this is a good idea, for the reasons already explained in the other issue.

**Comment by muradm** [comment]
@emersion, with this, if user explicitly requests the values of `WAYLAND_DISPLAY` and `DISPLAY` with explicit arguments, there is no need to check if there is a conflict. If user takes responsiblity to demand these values, user should take care of conflict situations or let it break/fail.

Otherwise, please show a way to run multiple independent wayland/sway on same machine.

**Comment by emersion** [comment]
Just read the values from Sway, instead of trying to force a specific one. The original issue even has example code for this.

**Comment by muradm** [comment]
The reason of closure of original issue states that: "...we can't easily make the difference between a broken...". This issue requests to not make any assumptions. If argument is explicitly specified with `--wayland-display display-1` and `--x-display :1`, on startup it is enough to see if `/run/user/<uid>/wayland-1` file and `/tmp/.X11-unix/X1` are present, regardless if they are working/broken and if present fail with meaningful error. Responsibility on providing valid values should be on caller.

After discussion on IRC, while script sample specified in closed issue or similar could be used, it has other drawbacks:

- default selection of wayland display and x display is unpredicatable, making automation around sway impossible
- additional control files as provided in the script example add additional fail points

The main goal of this issue, is to make sway more configurable by automation tools, which works best when values are predefined in advance.

Offered solution in closed issue could be usable by manually configured systems, in automated environments this brings a lot of uncertanity and additional possible fail points, for instance order of starting/restarting wayland/sway instances is one of them. Suppose there is a gpg-agent which requires x display value, when sway/xwayland is restarted it can acquire unpredictable value for x display, then all processes like gpg-agent in this case that require it will have to be restarted as well.

(Background context for my "automation", is that I use Guix and I'm strugling to port my xorg/i3 setup to wayland/sway due to this uncertanity in values of wayland display/x display).

**Comment by sorpaas** [comment]
I'm honestly facing the same issue. My use case is to allow multiple users sharing the same wayland socket. Right now, it requires specific ACL settings to `/run/user/<uid>` and `/run/user/<uid>/wayland-1`, which is really hack-y.

With a `--wayland-display` parameter, I can create a group `wayland`, a directory that all users in group `wayland` can access, and then create the wayland socket inside it. Then, once sway is started, I can su to any of the users in group `wayland` and start programs there.

**Comment by Masong19hippows** [comment]
@emersion It should not be the case that your script to execute sway should set environment variables for other applications to read. This is a tedious workaround for something sway should inherently do. Imagine if this was the case for every single other application on your computer - literally hundreds of custom scripts for no exact reason. Why should the user have to create a script for an inherit function of the protocol that the application is using? That's what the application is supposed to do/solve. What about cases where an app is being executed outside of any sourced profile - ex: systemd. You would then have to create a  script to edit the profile that the user loads. 

 Like @muradm said, this is something that the user can/should take care of if manually specified.

This terrible approach to an issue has definitely make the decision for me to not use sway.

**Comment by vintozver** [comment]
I believe one of two things have to happen:
1. allow sway to instantiate with the custom WAYLAND_DISPLAY & DISPLAY (X) values (with possibly error if the displays are already up and running by another display manager)
2. have a clear interface to communicate from sway - which WAYLAND_DISPLAY & DISPLAY (X) were chosen after the startup (for example, via the IPC socket)

I think option 2 would make more sense, it would still allow to fetch the actual chosen WAYLAND_DISPLAY and DISPLAY (X), without adding extra complexity into the startup scripts, error handing etc.
